pgjdbc/src/main/java/org/postgresql/util/internal/LeakTracker.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2021, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util.internal;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Consumer;
+
+public class LeakTracker<T> {
+  private final ReferenceQueue<T> leakedResources = new ReferenceQueue<>();
+
+  /**
+   * We need to keep a strong reference to the {@link LeakTraceHandle} otherwise
+   * it won't be enqueued to the reference queue.
+   */
+  @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+  private final ConcurrentMap<LeakTraceHandle<T>, Boolean> resourcesInUse =
+      new ConcurrentHashMap<>();
```

Comment:
Is a `Set<LeakTraceHandle<T>>` what we really want?

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#newKeySet--

Reply:
I am still not sure if the map or set is better here.
For instance, `QueryExecutorImpl` uses map in `HashMap<PhantomReference<Portal>, String> openPortalMap`